--- serverloop.c.orig	2024-03-11 05:19:20
+++ serverloop.c	2026-05-17 00:00:00
@@ -322,20 +322,24 @@ process_buffered_input_packets(struct ssh *ssh)
 	ssh_dispatch_run_fatal(ssh, DISPATCH_NONBLOCK, NULL);
 }
 
-static void
+static int
 collect_children(struct ssh *ssh)
 {
 	pid_t pid;
-	int status;
+	int status, collected = 0;
 
 	if (child_terminated) {
 		debug("Received SIGCHLD.");
 		while ((pid = waitpid(-1, &status, WNOHANG)) > 0 ||
-		    (pid == -1 && errno == EINTR))
-			if (pid > 0)
+		    (pid == -1 && errno == EINTR)) {
+			if (pid > 0) {
+				collected = 1;
 				session_close_by_pid(ssh, pid, status);
+			}
+		}
 		child_terminated = 0;
 	}
+	return collected;
 }
 
 void
@@ -344,7 +348,7 @@ server_loop2(struct ssh *ssh, Authctxt *authctxt)
 {
 	struct pollfd *pfd = NULL;
 	u_int npfd_alloc = 0, npfd_active = 0;
-	int r, conn_in_ready, conn_out_ready;
+	int r, conn_in_ready, conn_out_ready, collected_children;
 	u_int connection_in, connection_out;
 	sigset_t bsigset, osigset;
 
@@ -376,7 +380,11 @@ server_loop2(struct ssh *ssh, Authctxt *authctxt)
 		 */
 		if (sigprocmask(SIG_BLOCK, &bsigset, &osigset) == -1)
 			error_f("bsigset sigprocmask: %s", strerror(errno));
-		collect_children(ssh);
+		collected_children = collect_children(ssh);
+		if (collected_children &&
+		    !ssh_packet_is_rekeying(ssh) &&
+		    ssh_packet_have_data_to_write(ssh))
+			process_output(ssh, connection_out);
 		wait_until_can_do_something(ssh, connection_in, connection_out,
 		    &pfd, &npfd_alloc, &npfd_active, &osigset,
 		    &conn_in_ready, &conn_out_ready);
@@ -394,10 +404,14 @@ server_loop2(struct ssh *ssh, Authctxt *authctxt)
 		/* A timeout may have triggered rekeying */
 		if ((r = ssh_packet_check_rekey(ssh)) != 0)
 			fatal_fr(r, "cannot start rekeying");
+		if (!ssh_packet_is_rekeying(ssh) &&
+		    ssh_packet_not_very_much_data_to_write(ssh) &&
+		    channel_output_poll(ssh))
+			process_output(ssh, connection_out);
 		if (conn_out_ready)
 			process_output(ssh, connection_out);
 	}
-	collect_children(ssh);
+	(void)collect_children(ssh);
 	free(pfd);
 
 	/* free all channels, no more reads and writes */
